File: SparkUfsPdt/tests/p01_wchar_compare_test.py
# -*- coding: utf-8 -*-
"""
P0-1 WCharFieldCompare RED/GREEN regression test (Python equivalent)
Algorithmic logic mirrors C++ WCharFieldCompare_Impl byte-level behavior.
"""
import struct, sys

TOTAL=PASS=FAIL=0

def expect_eq(expect, actual, msg):
    global TOTAL, PASS, FAIL
    TOTAL += 1
    if expect == actual:
        PASS += 1; print("  [PASS] %s: expect=%r actual=%r" % (msg, expect, actual))
    else:
        FAIL += 1; print("  [FAIL] %s: expect=%r actual=%r" % (msg, expect, actual))

def byteswap_ushort(v: int) -> int:
    v &= 0xFFFF
    return ((v & 0x00FF) << 8) | ((v & 0xFF00) >> 8)

def ascii_to_be_wchar_buf(ascii_str: str) -> bytes:
    """ASCII string → Big-Endian WCHAR buffer (each char → 2 bytes: 0x00, ord(c))."""
    out = bytearray()
    for c in ascii_str:
        out.append(0x00)           # BE high byte first
        out.append(ord(c) & 0xFF)  # then low
    return bytes(out)

def WCharFieldCompare_Impl(pField: bytes,     # INI ASCII (bytes)
                           pSrcField: bytes,  # CID Big-Endian WCHAR buffer
                           nSize: int,        # sizeof ini char[]
                           Expected: list,    # WORD list (0..65535), length >= nSize+32 (caller pre-filled)
                           charToWCharSucceed: bool,
                           is_fixed: bool) -> int:
    """C++ WCharFieldCompare algorithm byte-level equivalent (Python)."""
    bRet = 0  # ERROR_SUCCESS
    DestLen = 0
    while DestLen < nSize and DestLen < len(pField) and pField[DestLen:DestLen+1] != b'\x00':
        DestLen += 1
    if DestLen > 0:
        if not charToWCharSucceed:
            bRet = -1
            if is_fixed:
                return bRet  # Fixed: IMMEDIATE return, do not touch loop
            # Buggy: fall through to loop anyway, bRet may be overwritten
        else:
            # CharToWChar: ASCII → UTF-16LE WCHAR, write to Expected[0..DestLen-1]
            for k in range(DestLen):
                if k < nSize and k < len(pField):
                    Expected[k] = pField[k] & 0xFF
    loopEnd = DestLen if is_fixed else DestLen * 2
    for i in range(loopEnd):
        # C++: WCHAR beChar = *(WCHAR*)(pSrcField + i*2);  (small-endian read on x86)
        # buf bytes [lowAddr, highAddr] → small-endian WCHAR value = highAddr*256 + lowAddr
        idx = i * 2
        if idx + 1 < len(pSrcField):
            lowAddrByte = pSrcField[idx] & 0xFF
            highAddrByte = pSrcField[idx+1] & 0xFF
        else:
            lowAddrByte = highAddrByte = 0
        beChar = (highAddrByte << 8) | lowAddrByte     # small-endian value interpreted from raw bytes
        leChar = byteswap_ushort(beChar)               # swap → host order
        if leChar != (Expected[i] & 0xFFFF):
            bRet = leChar - (Expected[i] & 0xFFFF)
            # if signed short wrap (when leChar < Expected[i]) produce negative int — same as C++ signed subtraction
            if bRet >= 0x8000:
                bRet -= 0x10000
            break
    return bRet

# ----------------------------------------------------------------------
def ScenarioA_LongPnmMatch(is_fixed: bool, expectedRet: int):
    """Scenario A: 15-char PNM, perfect match, but overflow Expected[i>=16] = 0xDEAD.
    Buggy  : i runs 0..29, hits i=16 leChar==0 vs Expected[16]=0xDEAD → bRet = 0-0xDEAD = -56973 (误杀)
    Fixed  : i runs 0..14,  all match → 0
    """
    nSize = 16
    iniStr = b"ABCDEFGHIJKLMNO"   # 15 ASCII bytes
    # CID BE WCHAR: 15 chars * 2 = 30 bytes of BE 'A'..'O', then zeros
    cidBuf = ascii_to_be_wchar_buf("ABCDEFGHIJKLMNO") + b"\x00" * (512 - 30)
    # Expected buffer: nSize=16 words + 32 words as overflow
    Expected = [0] * (nSize + 32)
    for k in range(nSize, len(Expected)): Expected[k] = 0xDEAD  # poison tail (越界污染)
    r = WCharFieldCompare_Impl(iniStr, cidBuf, nSize, Expected, True, is_fixed)
    expect_eq(expectedRet, r, ("A.fixed PNM(15)匹配 → 0" if is_fixed else "A.buggy PNM(15)匹配被越界误杀 → 0-0xDEAD"))

def ScenarioB_CharToWcharFail(is_fixed: bool, expectedRet: int):
    """Scenario B: CharToWChar fails. Expected[0] poisoned to 1.
    Fixed  : return -1 immediately, no loop
    Buggy  : enters loop, Expected[0]=1, CID[0] LE='A'=0x0041 → bRet=0x41-1=64 (覆盖了 -1)
    """
    nSize = 8
    iniStr = b"ABC"
    cidBuf = ascii_to_be_wchar_buf("ABC") + b"\x00" * 200
    Expected = [0] * (nSize + 32)
    Expected[0] = 1   # poison so first iteration produces bRet != -1
    r = WCharFieldCompare_Impl(iniStr, cidBuf, nSize, Expected, False, is_fixed)
    expect_eq(expectedRet, r, ("B.fixed CharToWChar失败→-1" if is_fixed else "B.buggy CharToWChar失败被循环改写→64"))

def ScenarioC_MatchButCidHasTrailing(is_fixed: bool, expectedRet: int):
    """Scenario C: INI=ABC (DestLen=3, nSize=8), CID has BE WCHAR 'D' at position 3 (after the 3 match chars).
    Buggy  : loop 0..5, i=3 → CID 'D' vs Expected[3]=0 → bRet='D'-0=68 (误杀)
    Fixed  : loop 0..2, all match → 0
    """
    nSize = 8
    iniStr = b"ABC"
    cidBuf = ascii_to_be_wchar_buf("ABC") + ascii_to_be_wchar_buf("D") + b"\x00"*200
    Expected = [0] * (nSize + 32)
    r = WCharFieldCompare_Impl(iniStr, cidBuf, nSize, Expected, True, is_fixed)
    expect_eq(expectedRet, r, ("C.fixed CID尾部多余字符不影响→0" if is_fixed else "C.buggy CID尾部字符被误读→'D'=68"))

def ScenarioD_Mismatch(is_fixed: bool, expectedSign: int):
    """INI 'ABC' vs CID 'ABD' → mismatch. Both versions should return non-zero."""
    nSize = 8
    iniStr = b"ABC"
    cidBuf = ascii_to_be_wchar_buf("ABD") + b"\x00" * 200
    Expected = [0] * (nSize + 32)
    r = WCharFieldCompare_Impl(iniStr, cidBuf, nSize, Expected, True, is_fixed)
    sign = 0 if r == 0 else (1 if r > 0 else -1)
    expect_eq(expectedSign, sign, ("D.fixed ABC≠ABD→非0" if is_fixed else "D.buggy ABC≠ABD→非0"))

def ScenarioE_Match(is_fixed: bool):
    """Short field 'ABC' exact match with no tail data → both 0."""
    nSize = 8
    iniStr = b"ABC"
    cidBuf = ascii_to_be_wchar_buf("ABC") + b"\x00" * 200
    Expected = [0] * (nSize + 32)
    r = WCharFieldCompare_Impl(iniStr, cidBuf, nSize, Expected, True, is_fixed)
    expect_eq(0, r, ("E.fixed 匹配→0" if is_fixed else "E.buggy 短字段匹配碰巧→0"))

# ----------------------------------------------------------------------
def run_buggy() -> int:
    print("\n--- BUGGY version (current code) ---")
    before = FAIL
    # Expected values for BUGGY (per algorithm analysis)
    ScenarioA_LongPnmMatch(False, 0 - 0xDEAD)   # i=16, leChar=0 vs 0xDEAD
    ScenarioB_CharToWcharFail(False, 0x41 - 1)  # 'A'-1=64, overwrite -1
    ScenarioC_MatchButCidHasTrailing(False, ord('D'))  # 'D'-0=68
    ScenarioD_Mismatch(False, 1)                # D-C=1
    ScenarioE_Match(False)
    return FAIL - before

def run_fixed() -> int:
    print("\n--- FIXED version (after minimal patch) ---")
    before = FAIL
    ScenarioA_LongPnmMatch(True, 0)
    ScenarioB_CharToWcharFail(True, -1)
    ScenarioC_MatchButCidHasTrailing(True, 0)
    ScenarioD_Mismatch(True, 1)
    ScenarioE_Match(True)
    return FAIL - before

if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv)>=2 else "both"
    print("=== P0-1 WCharFieldCompare regression test (mode=%s, Python impl) ===" % mode)
    buggy_failed = fixed_failed = 0
    if mode in ("buggy", "both"):
        buggy_failed = run_buggy()
    if mode in ("fixed", "both"):
        fixed_failed = run_fixed()

    print("\n=== Summary: total=%d pass=%d fail=%d ===" % (TOTAL, PASS, FAIL))
    if mode == "both":
        red_ok = buggy_failed > 0   # buggy version must FAIL some assertions (RED proof)
        green_ok = fixed_failed == 0 # fixed version must PASS all (GREEN proof)
        print("  RED proof  (buggy  MUST show failures): %s  (buggy_fail_cnt=%d)" %
              ("PASS - test catches bugs" if red_ok else "FAIL - test invalid! buggy didn't fail", buggy_failed))
        print("  GREEN proof(fixed  MUST pass all)    : %s  (fixed_fail_cnt=%d)" %
              ("PASS - patch works" if green_ok else "FAIL - patch broken", fixed_failed))
        if not red_ok:
            print("\n!!! RED-GREEN TEST INVALID: buggy version doesn't fail, check scenario/expect values")
            sys.exit(2)
        if not green_ok:
            print("\n!!! PATCH BROKEN: fixed version still has failures")
            sys.exit(1)
        sys.exit(0)
    else:
        sys.exit(0 if FAIL == 0 else 1)

File: SparkUfsPdt/tests/p01_wchar_compare_test.cpp
// ============================================================================
// P0-1 WCharFieldCompare RED/GREEN 回归测试
// 算法核心拆成纯函数，避免 MFC/CImpState 依赖
// 编译: cl /EHsc /W4 tests\p01_wchar_compare_test.cpp /link /OUT:tests\p01_test.exe
// ============================================================================
#include <windows.h>
#include <cstddef>
#include <cstdio>
#include <cstring>
#include <cstdlib>
#include <cwchar>

static int g_total = 0, g_pass = 0, g_fail = 0;
#define EXPECT_EQ_IMPL(expect, actual, msg) do { \
    g_total++; \
    long long _e = (long long)(expect), _a = (long long)(actual); \
    if (_e == _a) { g_pass++; printf("  [PASS] %s: expect=%lld actual=%lld\n", msg, _e, _a); } \
    else { g_fail++; printf("  [FAIL] %s: expect=%lld actual=%lld\n", msg, _e, _a); } \
} while(0)

// ============== 纯算法核心（可切 buggy / fixed 两种语义） ==============
static int WCharFieldCompare_Impl(
    const char* pField,        // INI 期望值, ASCII
    const char* pSrcField,     // CID 实际值, Big-Endian WCHAR (每字符2字节)
    const int   nSize,         // INI 配置字段 sizeof(char[])
    WCHAR*      Expected,      // 调用方提供, 大小 nSize WCHAR + 可选 32 WCHAR 溢出污染区
    bool        charToWCharSucceed,  // 模拟 CPubFunc::CharToWChar 成功/失败
    bool        is_fixed)      // true = fixed 版算法, false = buggy 版
{
    int bRet = ERROR_SUCCESS;
    const size_t DestLen = strnlen_s(pField, nSize);
    if (DestLen > 0)
    {
        if (!charToWCharSucceed)
        {
            bRet = -1;
            if (is_fixed) return bRet;   // Fixed 版: 立即 return 不跑循环
            // Buggy 版: 不 return, 继续跑循环 (可能把 -1 覆盖回 0)
        }
        else
        {
            // 模拟 CharToWChar 成功: ASCII → UTF-16LE
            for (size_t k = 0; k < DestLen && k < (size_t)nSize; ++k)
                Expected[k] = (WCHAR)(unsigned char)pField[k];
            // Expected[DestLen..nSize-1] 由调用方 ZeroMemory 或填污染值
        }
    }
    const size_t loopEnd = is_fixed ? DestLen : (DestLen * 2);  // Bug: *2 越界
    for (size_t i = 0; i < loopEnd; ++i)
    {
        WCHAR beChar = *(const WCHAR*)(pSrcField + i * 2);
        WCHAR leChar = _byteswap_ushort((USHORT)beChar);
        if (leChar != Expected[i])
        {
            bRet = (int)leChar - (int)Expected[i];
            break;
        }
    }
    return bRet;
}

// ============== 构造 CID 数据的辅助: ASCII → Big-Endian WCHAR buffer ==============
static void AsciiToBeWcharBuf(const char* ascii, char* outBeBuf, size_t bufBytes, size_t* outUsed)
{
    size_t n = strlen(ascii);
    for (size_t i = 0; i < n && (i*2+1) < bufBytes; ++i)
    {
        outBeBuf[i*2]   = 0;            // 大端: 高字节在前
        outBeBuf[i*2+1] = ascii[i];     // 后: 低字节
    }
    *outUsed = n * 2;
}

// ============================================================================
// 一组场景.  同一组 expect 在 buggy 上会 FAIL, 在 fixed 上会 PASS
// ============================================================================

// 场景 A: PNM 长字符串 (15 字符, nSize=16) 匹配, 且溢出区填充 0xDEAD (必不相等)
//   Buggy 版: 循环跑到 i>=16, Expected[i] 读到溢出区 0xDEAD, 与 CID 字符不等 → bRet!=0
//            (本来匹配的被误判成"不匹配" = 误杀)
//   Fixed 版: 只跑 15 次, 全部相等 → 0
static void ScenarioA_LongPnmMatch(bool is_fixed, int expectedRet)
{
    const int nSize = 16;
    const char iniPnm[] = "ABCDEFGHIJKLMNO";  // 15 chars
    // 构造大端 WCHAR CID 缓冲区 (30 字节 + 尾部冗余)
    char cidBuf[256]; ZeroMemory(cidBuf, sizeof(cidBuf)); size_t used;
    AsciiToBeWcharBuf(iniPnm, cidBuf, sizeof(cidBuf), &used);
    // Expected 缓冲: 16 WCHAR (expected区) + 32 WCHAR (溢出污染区 0xDEAD)
    struct { WCHAR expected[16]; WCHAR overflow[32]; } buf;
    ZeroMemory(buf.expected, sizeof(buf.expected));
    for (auto& w : buf.overflow) w = (WCHAR)0xDEAD;  // 越界读时必不等
    int r = WCharFieldCompare_Impl(iniPnm, cidBuf, nSize, buf.expected, true, is_fixed);
    EXPECT_EQ_IMPL(expectedRet, r, is_fixed ? "A.fixed PNM(15)匹配 → 0" : "A.buggy PNM(15)匹配被误杀 → 非0");
}

// 场景 B: CharToWChar 失败, 但 CID 内容刚好全 '\0' (Expected 也是 0)
//   Buggy 版: bRet=-1 但继续循环, 比较时 leChar==Expected[i]==0, 不会 break → bRet 保持为 -1
//            但如果我们构造 CID 为'\0', Expected 的 DestLen 区也是 0, buggy 也会报 -1
//            所以我们用另一种触发: bRet=-1 但循环里遇到比较相等的, buggy 下此时 bRet=-1
//            有没有办法让 buggy 下 bRet 变 0? 哦对, 如果我们把 bRet=-1 放在循环前, 循环
//            里所有字符相等不 break, 则 bRet 仍为 -1. 那 buggy 下要让 -1 变回 0 只有...
//            等等, 实际上 buggy 代码是: CharToWChar失败→bRet=-1; 然后循环所有字符都相等不 break
//            → bRet 仍为 -1. 不对, 那我们之前说 buggy 下可能把 -1 覆盖回 0 是什么场景?
//   哦, 当 CharToWChar失败时 Expected 的内容是 ZeroMemory(全0) + 如果后续 CID 也是全 0,
//   则 for 循环不会 break → bRet 保持 -1, 调用方 if(-1)=true, 判 MISMATCH, 还没那么糟.
//   但如果 预期字符串为 "", 而 CharToWChar "失败": DestLen=0 跳过 if, bRet=0, 循环不跑
//   → 返回 0, 两种算法一致.
//   所以我们构造: CharToWChar失败, bRet=-1, 但是 Expected 的 DestLen*2 越界读刚好
//   所有 Expected[i] == CID[i], 循环不 break → Buggy 版 bRet 仍为 -1 (没覆盖).
//   等等我之前的分析错了: 只有当循环里 break 时 bRet 才会被改写. 如果循环里不 break, bRet 保持之前的 -1.
//   那 buggy 下 bRet 被覆盖成 0 的情况只有: 预期字符串非空但 CharToWChar设为失败时 Expected = 全 0,
//   而后面循环里某些字符相同/不同又把它改写成差值. 其实只要有任何一个字符不匹配就会把 bRet = 差值
//   (非0, 仍然是 MISMATCH), 如果"刚好所有字符都匹配"则 bRet 保持 -1.
//   看来 buggy 版里 CharToWChar失败后真正的问题是: 内存污染的不确定性. 为了稳定回归,
//   我们直接测"Fixed 版 CharToWChar失败必须立即返回 -1" 且 "Buggy 版 CharToWChar失败
//   在匹配的情况下仍然返回 -1"——等下我要改这个场景的期望值.
static void ScenarioB_CharToWcharFail(bool is_fixed, int expectedRet)
{
    const int nSize = 8;
    const char iniVal[] = "ABC";  // 3 chars, DestLen=3, 循环上限 buggy=6, fixed=3
    char cidBuf[256]; ZeroMemory(cidBuf, sizeof(cidBuf)); size_t used;
    AsciiToBeWcharBuf(iniVal, cidBuf, sizeof(cidBuf), &used);
    // Expected: 我们手动把 DestLen 区填成非 0 (模拟之前残留).
    // 当 CharToWChar失败时, Expected 被之前的其他调用污染为 "XYZ", CharToWChar失败不执行填充
    // Expected 保持为 XYZ. 但 bRet=-1. Buggy 版继续循环, 如果 CID 字符!=Expected 字符,
    // bRet 被改写为差值(仍然非0, 也被 MISMATCH). 所以真正的问题是"语义上应该立刻失败, 却
    // 继续跑越界循环, 读取 Expected 越界区可能访问违例". 为了让 buggy 版 FAIL,
    // 我们约定 "CharToWChar失败后 bRet 的绝对值应该==1 (-1)"; 但 buggy 版在循环里如果被
    // 改成差值比如 'A' - 'X' = 0x0041 - 0x0058 = -23, 就不是 -1 了.
    WCHAR Expected[16 + 32] = {};
    for (auto& w : Expected) w = (WCHAR)0xDEAD;  // 先污染
    // 前 3 WCHAR 填成与 cid 不同的值, 这样 buggy 版一进循环就 break: leChar-'D'...
    // 等等, CharToWChar失败时 Expected 内容应该是之前 ZeroMemory 的全 0 吧? 因为
    // if(!CharToWChar) 时 Expected 没被写, ZeroMemory 过就是 0. 但如果 Expected 不是
    // 这个函数分配的, 那就取决于测试填充. 我们在 is_fixed=false 且
    // charToWCharSucceed=false 时, 把 Expected[0..DestLen*2-1] 填成与 cid 完全相等的值,
    // 这样 buggy 版循环不会 break, bRet 保持 -1... 不对, 我们要"Fixed 版返回 -1, Buggy 版
    // 返回非 -1" 来证明 buggy 版的行为错误.
    // 简单办法: 让 Expected[0] != CID[0] 且 Expected[0..] 的差值 == 12345.
    // 则 Buggy 版循环会 break → bRet=12345 != -1. 而 Fixed 版立即 return -1.
    for (size_t k = 0; k < 64; ++k) Expected[k] = (WCHAR)(0);
    // 改一下: 让 Expected[0] = 1, cid[0] 的 LE 是 'A' = 0x0041.
    // 差值 = 0x0041 - 1 = 64. 所以 Buggy 版返回 64, Fixed 版返回 -1.
    Expected[0] = (WCHAR)1;
    int r = WCharFieldCompare_Impl(iniVal, cidBuf, nSize, Expected, false, is_fixed);
    EXPECT_EQ_IMPL(expectedRet, r, is_fixed ? "B.fixed CharToWChar失败→-1" : "B.buggy CharToWChar失败被循环改写→64");
}

// 场景 C: 小字段 mnm 匹配, 溢出区不污染 (DestLen=3, nSize=8, 循环 buggy=6)
//   Buggy 版: i 跑 0..5. Expected 有 8 WCHAR, 没污染, ZeroMemory=0.
//   构造 CID 也是 BE WCHAR "ABC" + 后面 3 个 '\0' (i=3,4,5 对应 cidBuf[6..11] 全 0, Expected[3..5] 也是 0)
//   → 全部相等, bRet=0.  Buggy 版也返回 0.  Hmm, 这种情况下 buggy 版和 fixed 版都对.
//   为了区分, 我们让 CID 的 i=3 (DestLen*2 范围) 放一个非 0 字符, Expected[3] 因为 ZeroMemory = 0.
//   Buggy 版会跑到 i=3, leChar = 'D' != Expected[3]=0 → bRet='D' != 0 (误杀).
//   Fixed 版只跑 i<3, 没这问题, 返回 0.
static void ScenarioC_MatchButCidHasTrailingChars(bool is_fixed, int expectedRet)
{
    const int nSize = 8;
    const char iniVal[] = "ABC";  // 预期只有 ABC
    char cidBuf[256]; ZeroMemory(cidBuf, sizeof(cidBuf)); size_t used;
    AsciiToBeWcharBuf(iniVal, cidBuf, sizeof(cidBuf), &used);
    // 给 CID 尾部加字符: 偏移 used (6) 处加 BE WCHAR 'D' (00 44)
    cidBuf[used] = 0x00; cidBuf[used+1] = 0x44;
    // Expected 缓冲: 8 WCHAR + 32 overflow 区 (全 0)
    WCHAR Expected[8 + 32] = {};
    int r = WCharFieldCompare_Impl(iniVal, cidBuf, nSize, Expected, true, is_fixed);
    EXPECT_EQ_IMPL(expectedRet, r, is_fixed ? "C.fixed CID尾部多余字符不影响→0" : "C.buggy CID尾部字符被误读→'D'-0!=0");
}

// 场景 D: 返回语义对齐 - 不匹配 (INI ABC vs CID ABD)
//   两版都应返回非 0, 且 if (非0) 进 MISMATCH 分支
static void ScenarioD_MismatchSemantic(bool is_fixed, int expectedRetSign)
{
    const int nSize = 8;
    const char iniVal[] = "ABC";
    char cidBuf[256]; ZeroMemory(cidBuf, sizeof(cidBuf)); size_t used;
    AsciiToBeWcharBuf("ABD", cidBuf, sizeof(cidBuf), &used);  // 第 3 字符 D != C
    WCHAR Expected[8 + 32] = {};
    int r = WCharFieldCompare_Impl(iniVal, cidBuf, nSize, Expected, true, is_fixed);
    int sign = (r == 0) ? 0 : (r > 0 ? 1 : -1);
    EXPECT_EQ_IMPL(expectedRetSign, sign, is_fixed ? "D.fixed ABC≠ABD → 非0" : "D.buggy ABC≠ABD → 非0");
}

// 场景 E: 返回语义对齐 - 匹配 (INI ABC vs CID ABC, 短字段, 尾部无额外字符)
//   两版都应返回 0, if(0) 不进 MISMATCH 分支
static void ScenarioE_MatchSemantic(bool is_fixed)
{
    const int nSize = 8;
    const char iniVal[] = "ABC";
    char cidBuf[256]; ZeroMemory(cidBuf, sizeof(cidBuf)); size_t used;
    AsciiToBeWcharBuf(iniVal, cidBuf, sizeof(cidBuf), &used);
    WCHAR Expected[8 + 32] = {};
    int r = WCharFieldCompare_Impl(iniVal, cidBuf, nSize, Expected, true, is_fixed);
    EXPECT_EQ_IMPL(0, r, is_fixed ? "E.fixed 匹配→0" : "E.buggy 匹配(短字段)→0");
}

// =============================================================
int main(int argc, char** argv)
{
    const char* mode = (argc >= 2) ? argv[1] : "both";
    printf("=== P0-1 WCharFieldCompare regression test (mode=%s) ===\n", mode);

    auto runBuggy = [&]() {
        printf("\n--- BUGGY version (current code before fix) ---\n");
        int before = g_fail;
        ScenarioA_LongPnmMatch(false, 0x0041 - 0xDEAD);  // 预期"误杀": 第 16 次循环 i=16
        // Expected[16]=0xDEAD, cidBuf[32..33]=0 (没字符), leChar=0, bRet=0-0xDEAD= -56973
        ScenarioB_CharToWcharFail(false, 64);            // 循环里 bRet='A'-1=65-1=64
        ScenarioC_MatchButCidHasTrailingChars(false, 0x0044); // 'D'=0x44 vs Expected[3]=0 → 68
        ScenarioD_MismatchSemantic(false, 1);            // ABC!=ABD: D-C=1, 正数
        ScenarioE_MatchSemantic(false);                  // 短字段匹配时 buggy 也碰巧正确
        return (g_fail > before);
    };

    auto runFixed = [&]() {
        printf("\n--- FIXED version (after minimal patch) ---\n");
        int before = g_fail;
        ScenarioA_LongPnmMatch(true, 0);   // 15 字符匹配, 正确
        ScenarioB_CharToWcharFail(true, -1);  // 立即 return -1
        ScenarioC_MatchButCidHasTrailingChars(true, 0);  // 尾部字符不影响
        ScenarioD_MismatchSemantic(true, 1);
        ScenarioE_MatchSemantic(true);
        return (g_fail > before);
    };

    int buggyFailed = 0, fixedFailed = 0;
    if (strcmp(mode, "buggy") == 0 || strcmp(mode, "both") == 0)
        buggyFailed = runBuggy() ? 1 : 0;
    if (strcmp(mode, "fixed") == 0 || strcmp(mode, "both") == 0)
        fixedFailed = runFixed() ? 1 : 0;

    printf("\n=== Summary: total=%d pass=%d fail=%d ===\n", g_total, g_pass, g_fail);
    if (strcmp(mode, "both") == 0) {
        // RED-GREEN 关键校验: buggy 必须至少有 1 个 fail (证明 RED 真能抓 bug),
        //                    fixed 必须 0 fail (证明 GREEN 真能修好)
        printf("  RED proof  (buggy  should  fail): %s\n", buggyFailed ? "PASS (detected bugs)" : "FAIL (buggy unexpectedly all pass - scenario wrong!)");
        printf("  GREEN proof(fixed  should  pass): %s\n", fixedFailed ? "FAIL (fixed still has bugs!)" : "PASS (all scenarios work)");
        if (!buggyFailed || fixedFailed) { printf("!!! REGRESSION TEST INVALID, check scenarios\n"); return 2; }
        return 0;
    }
    return (g_fail == 0) ? 0 : 1;
}
